internal/unionstore/art.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 TiKV Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:unused
+package unionstore
+
+import (
+	"math"
+	"sync"
+
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/kv"
+)
+
+type ART struct {
+	sync.RWMutex
```

Comment:
What are the fields that should be protected by the mutex?

Reply:
It's required by the membuffer interface.

https://github.com/tikv/client-go/blob/75e3705e58f1c9f68cce3467de4429ea959b59b7/internal/unionstore/union_store.go#L171-L174

In TiDB, some SQL executors will read and write membuffer concurrently, the membuffer needs to be guarded by mutex in this case.

Reply:
As it's a memdb requirement/constraint instead of an `ART` one, how about
```
type ARTMemDB struct {
     sync.RWMutex
     artImpl  ART // ART is a single threaded data strcuture implementation.
}
```